### DISCUSSION TITLE
On how to use A2DP sink along with ESP32-audioI2S, sharing ressources.

### DISCUSSION THREAD

**Discussion by 0x0fe** [schreibfaul1/ESP32-audioI2S#548] [Q&A]
So i use ESP32-audioI2S for all functions related SD_MMC player,ne audio streaming, and web audio streaming + downloading. But there is another source which i want to use : Bluetooth A2DP, which is not integrated into ESP32-audioI2S.

What a wonder is how i could share some of the ressources (i think of the main ringbuffer) between audioI2S and the A2dP sink;
he A2Dp sink is part of idF libraries, and i use the excellent wrapper from [pschatzmann](https://github.com/pschatzmann/ESP32-A2DP/commits?author=pschatzmann) 

Or, if not possible, how i can free up Audio I2S ressource when A2DP is started and reinstate them afterwards.
Because of course, audioI2S and A2DP sink are used exclusively and never simultaneously.

I am aware of the pull request for A2DP source, but the probelmatic here was not to share ressource but rather to feed the A2DP source from audioI2S. I the case of A2DP sink there probably no need to feed A2DP datas into audioI2S, unless it is simpelr than sharing ressources liek the ringbuffer.

**Comment by schreibfaul1** [comment]
I don't have an executable program for this. If you are interested in BT, I would like to point you to the ESPuino project, where audioI2S works together with BT (sink and source).
https://github.com/biologist79/ESPuino
https://forum.espuino.de/

**Reply by 0x0fe** [reply]
Yes, in fact i checked it quickly when i found the pull request for a2dp source, for sink it seems they used the [pschatzmann](https://github.com/pschatzmann/ESP32-A2DP/commits?author=pschatzmann) a2dp sink wrapper too. 

For now i am trying to free up ressources on audioI2S before starting A2DP sink, and free up A2DP sink ressource before reinitilizing audioI2S.
I'll report if it works fine, but my concern is that i did not find a way to defer the audioI2S initialisation, it is all done at instanciation automatically.

```cpp
void start_player(void){

  //defered initialisation
  audioTask=xTaskCreatePinnedToCore(audio_task,"audioplay",5000,NULL,2|portPRIVILEGE_BIT,NULL,1);
}
void stop_player(void){

  vTaskDelete(audioTask);
  free(audio);
}
```

```cpp
void start_a2dp(void){

  a2dp_sink.start("A2DP");
}
void stop_a2dp(void){
  
  a2dp_sink.end(true);
}
```

**Reply by 0x0fe** [reply]
so, this leads to an error when we try to use the I2S elsewhere after it has been freed in A2DP sink,  a ticket is open for that : https://github.com/espressif/esp-idf/issues/11642
